# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [1]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [2]:
! pip install -q uncertainties

To install the module before you can import it:

In [3]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [4]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [5]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [6]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [7]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

np.float64(0.01739540169125163)

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [8]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [26]:
# average velocity of an object and its error
x = uc.ufloat(5.1, 0.4) # m, distance that the object travels with its corresponding error.
t = uc.ufloat(0.4, 0.1) # s, time in which x is travelled, with its error.
v = x/t # the average velocity is the distance diveded by time
print('The average velocity of the obejct is', v, 'm/s')

The average velocity of the obejct is 12.7+/-3.3 m/s


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [25]:
# calculate the uncertainty of h
g = 9.81 # m/s^2, the acceleration due to gravity
v0 = uc.ufloat (4.0, 0.2) # m/s, the initial speed of the cow
t = uc.ufloat (0.60, 0.06) # s, the time after which h equals
h = v0*t - 1/2 * g* t**2
print('The height of the cow is', h, 'm')

The height of the cow is 0.63+/-0.16 m


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [23]:
# find local length f of a lens from the object distance and the image distance, with its correponding error
u = uc.ufloat(20, 0.5) # cm, the object distance with its corresponding error
v = uc.ufloat(10, 0.5) # cm, the image distance with its corresponding error
f = 1/(1/u + 1/v)
print('The focal length of the lens is', f, 'cm')

The focal length of the lens is 6.67+/-0.23 cm,


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [40]:
# agreement in this case means that both share uncertainty ranges
jack = uc.ufloat(1.33, 0.03) # the refractive index of water measured by Jack with its corresponding error
jill = uc.ufloat(1.28, 0.02) # the refractive index of water measured by Jill with its correponding error

# check overlapping of their uncertainty ranges
if (jack.nominal_value - jack.std_dev <= jill.nominal_value + jill.std_dev) and \
   (jack.nominal_value + jack.std_dev >= jill.nominal_value - jill.std_dev):
    print("The two measurements are in agreement.") # as jack's smaller value equals jill's higher, their values are in agreement
else:
    print("The two measurements are not in agreement.")

The two measurements are in agreement.


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [21]:
# finding the natural frequency and its corresponding error
f_res = uc.ufloat(23.2, 0.1) # Hz, the measured resonance frequency with its corresponding error
omega_res = f_res * 2 * np.pi # converting the resonance frequency to angular resonance frequency
alpha = uc.ufloat(19.5, 0.5) # s^-1, the measured damping coefficient with its corresponding error
omega_0 = um.sqrt(omega_res**2 + 2*alpha**2)
f_0 = omega_0/(2*np.pi) # transforming natural angular frequency to angular frequency.
print('The natural frequency of the system is', f_0, 'Hz')

The natural frequency of the system is 23.61+/-0.10 Hz


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [20]:
# calculate the resonant frequency of sound waves and its error for (nx, ny, nz) = (1,1,1)
l_x = uc.ufloat (10.2, 0.2) # m
l_y = uc.ufloat (5.2, 0.3) # m
l_z = uc.ufloat (20.0, 0.1) # m
T = uc.ufloat (23, 1) # C
c = 331.3 + T*0.606 # m/s
n_x_1 = 1
n_y_1 = 1
n_z_1 = 1
f_1 = (c/2)* (um.sqrt((n_x_1**2/l_x**2)+(n_y_1**2/l_y**2)+(n_z_1**2/l_z**2)))
print('The first resonant frequency of sound waves is', f_1, 'Hz')

# calculate f and its error for (nx, ny, nz) = (1,1,2)
l_x = uc.ufloat (10.2, 0.2) # m
l_y = uc.ufloat (5.2, 0.3) # m
l_z = uc.ufloat (20.0, 0.1) # m
T = uc.ufloat (23, 1) # C
c = 331.3 + T*0.606 # m/s
n_x_2 = 1
n_y_2 = 1
n_z_2 = 2
f_2 = (c/2)* (um.sqrt((n_x_2**2/l_x**2)+(n_y_2**2/l_y**2)+(n_z_2**2/l_z**2)))
print('The second resonant frequency of sound waves is', f_2, 'Hz')

# calculate f and its error for (nx, ny, nz) = (2,1,1)
l_x = uc.ufloat (10.2, 0.2) # m
l_y = uc.ufloat (5.2, 0.3) # m
l_z = uc.ufloat (20.0, 0.1) # m
T = uc.ufloat (23, 1) # C
c = 331.3 + T*0.606 # m/s
n_x_3 = 2
n_y_3 = 1
n_z_3 = 1
f_3 = (c/2)* (um.sqrt((n_x_3**2/l_x**2)+(n_y_3**2/l_y**2)+(n_z_3**2/l_z**2)))
print('The third resonant frequency of sound waves is', f_3, 'Hz')

The first resonant frequency of sound waves is 38.2+/-1.7 Hz
The second resonant frequency of sound waves is 41.1+/-1.6 Hz
The third resonant frequency of sound waves is 48.2+/-1.4 Hz


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [19]:
# calculate the reflection coefficient's error
theta_i_degrees = uc.ufloat (78, 1) # degrees
theta_t_degrees = uc.ufloat (40, 1) # degrees
theta_i = theta_i_degrees * np.pi/180 # converting degrees to radians
theta_t = theta_t_degrees * np.pi/180
R_parallel = (um.tan(theta_i - theta_t)**2)/(um.tan(theta_i + theta_t)**2)
print(' The reflection coefficient is', R_parallel,)

 The reflection coefficient is 0.173+/-0.027


## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [36]:
# calculate the fractional uncertainty as a percentage to one signigicant figure
dot_m = uc.ufloat(0.13, 0.01)  # kg/s, the mass of flow rate
d = uc.ufloat(11, 1)  # mm, the orifice diameter
D = uc.ufloat(71, 1)  # mm, the pipe diameter
rho = uc.ufloat(1.01, 0.01)  # g/cm^3, the fluid density
delta_P = uc.ufloat(156, 7)  # Pa, the differential pressure
K = 1  # Placeholder value for K
F = 1  # Placeholder value for F

# convert mm to meter for both d and D
d_m = d / 1000
D_m = D / 1000

# convert g/cm^3 to kg/m^3 for rho (from g to kg is /1000, but from /cm^3 to /m^3 we multiply by 1000000, then in the end just multiply 1000)
rho_kg_m3 = rho * 1000

# fluid flow discharge coefficient
C_d = (dot_m * um.sqrt(1 - (d_m/D_m)**4)) / (K * d_m**2 * F * um.sqrt(rho_kg_m3 * delta_P))

# calculate the fractional uncertainty of C_d
fractional_uncertainty = C_d.std_dev / C_d.nominal_value

# convert to percentage
fractional_uncertainty_percentage = fractional_uncertainty * 100

print (f'The fractional uncertainty is {fractional_uncertainty_percentage:.1f}%')


The fractional uncertainty is 19.9%


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.